File: ctci/java/ch1-arrays-and-strings/q7_rotate_matrix/Question.java
package q7_rotate_matrix;

/**
 * Created by jjlee on 7/12/16.
 */
public class Question {
    /* in place using temp variable and layer-based iteration */
    public static void rotate(int[][] matrix) {
        int len = matrix.length;
        for (int i = 0; i < len / 2; i++) {
            for (int j = i; j < len - 1 - i; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[len - 1 - j][i];
                matrix[len - 1 - j][i] = matrix[len - 1 - i][len - 1 - j];
                matrix[len - 1 - i][len - 1 - j] = matrix[j][len - 1 - i];
                matrix[j][len - 1 - i] = temp;
            }
        }
    }
}

File: ctci/python/ch1-arrays-and-strings/q7-rotate-matrix/Question.py
# in place using temp variable and layer-based iteration
def rotate(matrix):
    length = len(matrix)
    for i in xrange(length / 2):
        for j in xrange(i, length - 1 - i):
            temp = matrix[i][j]
            matrix[i][j] = matrix[length - 1 - j][i]
            matrix[length - 1 - j][i] = matrix[length - 1 - i][length - 1 - j]
            matrix[length - 1 - i][length - 1 - j] = matrix[j][length - 1 - i]
            matrix[j][length - 1 - i] = temp
